text/0000-newtype.md

```diff
@@ -0,0 +1,360 @@
+with `as` and create a lot of friction in the code:
+
+```rust
+fn gen(input: i64) -> i32 {
+    unsafe { Gen(FromPrimitive::from_i64(input).unwrap()).0 as i32 }
+}
+```
+
+(Note that the error checking performed by `FromPrimitive` above is not always
+desired.)
+
+Below we propose an alternative that sits between `type` and newtype structs.
+
+# Detailed design
+
+Rename `type` to `alias` to better document its function.
+
+Add a `newtype` keyword.
+
+## Syntax
+
+```rust
+newtype T = U;
+```
+
+where `T` and `U` are as in the definition of `alias`.
+
+(You might want to first skip the details and continue with the "Rationale"
+section below.)
+
+## Semantics
+
+`T` behaves as if it had been declared by `struct T(());` except as described
+below.
+
+### Representation
+
+`T` has the same representation in memory as `U` if any.
+
+### Casting
+
+Let `R1` be the binary, transitive, symmetric, irreflexive relation generated by
+all `newtype` definitions, all `alias` definitions, and the numeric casts built
+into the language. 
+
+If `(T, U) \in R1` and `x: T`, then `x as U` is a valid cast.
+
+Let `R2` be the binary, transitive, symmetric, irreflexive relation generated by
+all `newtype` definitions and all `alias` definitions.
+
+Let `S` be the set of reference operators, `S = {&, &mut}`, and let `W` be a
+word generated by `S`, e.g., `W = e` or `W = &&&mut &&mut &`.
+
+If `(T, U) \in R2` and `x: WT`, then `x as WU` is a valid cast.
+
+### Inference
+
+Let `R2` be as above.
+
+If `U` is an integer type and `(T, U) \in R2`, then untyped integer literals will
+be infered as `T` in `T` position.
+
+If `U` is a floating point type and `(T, U) \in R2`, then untyped floating point
+literals will be infered as `T` in `T` position.
+
+### Traits
+
+No built-in traits are or can be implemented for `T` except as described below.
+
+For all types `W`, `PartialOrd<W>` can be implemented for `T`. If
+`PartialOrd<T>` is not explicitly implemented for `T` and `PartialOrd<U>` is
+implemented for `U`, then `PartialOrd<T>` shall automatically be implemented for
+`T` via casting to `U`.
+
+`Neg` can be implemented for `T`. If `Neg` is not explicitly implemented for `T`
+and `Neg` is implemented for `U`, then `Neg` shall automatically be implemented
+for `T` via casting to `U`.
+
+For all types `W`, `Shl<W>` can be implemented for `T`. If for any integer type
+`I` `Shl<I>` is not is not explicitly implemented for `T` and `Shl<I>` is
+implemented for `U`, then `Shl<I>` shall automatically be implemented for `T`
+via casting to `U`.
+
+`T` implements `Drop` if and only if `U` implements `Drop`.
+
+`T` implements `Send` if and only if `U` implements `Send`.
+
+`T` implements `Sized` if and only if `U` implements `Sized`.
+
+`T` implements `Copy` if and only if `U` implements `Copy`.
+
+`T` implements `Sync` if and only if `U` implements `Sync`.
+
+For all types `W`, `Fn<W>` can be implemented for `T`.
```

Comment:
It would be easier to say `Index, IndexMut, Deref, and DerefMut can be implemented or not implemented as usual.`